### PR TITLE
enhancement(vrl): `parse_key_value` fixes

### LINKED ISSUE #7839: `parse_key_value` not correctly parsing quoted value
Reported by user here: https://github.com/timberio/vector/discussions/7772#discussioncomment-857602

Vector version:

```
vector 0.14.0 (x86_64-apple-darwin 5f3a319 2021-06-03)
```

It looks like `parse_key_value`, in this particular case, is not correctly parsing a quoted field value.

```
$ parse_key_value!("time=\"2021.2021-06-11 09:54:11Z\", severity=[php7:emerg] , process=1700689 , message=\"PHP Parse error:  syntax error, unexpected '->' (T_OBJECT_OPERATOR) in /var/www/prod/releases/master-c7225365fd9faa26262cffeeb57b31bd7448c94a/source/index.php on line 14\"")
{ ", message": "PHP Parse error:  syntax error, unexpected '->' (T_OBJECT_OPERATOR) in /var/www/prod/releases/master-c7225365fd9faa26262cffeeb57b31bd7448c94a/source/index.php on line 14", ", process": "1700689", "09:54:11Z", severity": "[php7:emerg]", "time": "\"2021.2021-06-11" }
```

Here the `time` key didn't read the whole value even though it is quoted.

### PR DESCRIPTION
**Pull request by prognant**
## Attempt to fix two issues:
* Escaping ( closes #7839 ) problem, e.g.:
```
$ parse_key_value!(s'user="DOMAIN\account space" ip="1.2.3.4 space"', whitespace:"strict")
{ "space" ip": "1.2.3.4 space", "user": "\"DOMAIN\\account" }
```

* Quoted standalone key value that include the field separator, e.g.:
```
$ parse_key_value!(s'"foo==b" bar=baz', whitespace:"strict")
{ ""foo": "=b\"", "bar": "baz" }
```
All older tests are passing unmodified, but I'll probably add more to better cover corner cases.

## Additional issue
Will create an issue about escaping because this:
```
$ parse_key_value!(s'foo="\"bar\""', whitespace:"strict")
{ "foo": "\\\"bar\\\"" }
```
should not happen, we should get (to be symmetric W.R.T `encode_key_value`):
```
{ "foo": "\"bar\"" }
```

### CODE REVIEW

File: lib/vrl/stdlib/src/parse_key_value.rs [hunk 1]
```diff
@@ -373,14 +375,23 @@ fn parse_value<'a, E: ParseError<&'a str> + ContextError<&'a str>>(
...
         alt((
             parse_delimited('"', key_value_delimiter),
-            parse_undelimited(key_value_delimiter),
+            verify(parse_delimited('"', field_delimiter), |_: &str| {
+                standalone_key
+            }),
+            verify(parse_undelimited(key_value_delimiter), |s: &str| {
+                !standalone_key || !s.contains(field_delimiter)
+            }),
+            parse_undelimited(field_delimiter),
```
Comment (prognant) [inline lib/vrl/stdlib/src/parse_key_value.rs:394]: I'll switch that conditional to https://docs.rs/nom/6.1.2/nom/combinator/fn.cond.html

File: lib/vrl/stdlib/src/parse_key_value.rs [hunk 2]
```diff
@@ -373,22 +374,51 @@ fn parse_value<'a, E: ParseError<&'a str> + ContextError<&'a str>>(
 }
 
 /// Parses the key.
-/// Parsing strategies are the same as parse_value, but we don't need to convert the result to a `Value`.
+/// Overall parsing strategies are the same as parse_value, but we don't need to convert the result to a `Value`.
+/// Standalone key are handle here so quoted standalone key that contains a delimiter will be dealt with correctly.
```
Comment (spencergilbert) [inline lib/vrl/stdlib/src/parse_key_value.rs:378]:
```suggestion
/// Standalone key are handled here so a quoted standalone key that contains a delimiter will be dealt with correctly.
```
nit

**Review by spencergilbert — APPROVED** [review]

**Review by jdrouet — APPROVED** [review]
LGTM
